service/switchboardapp/src/test/java/com/switchboard/app/DecoderControllerTest.java

```diff
@@ -0,0 +1,112 @@
+package com.switchboard.app;
+
+import com.switchboard.app.controller.DecoderController;
+import com.switchboard.app.dao.DecoderDaoImpl;
+import com.switchboard.app.dao.DeviceDaoImpl;
+import com.switchboard.app.domain.DecoderEntity;
+import com.switchboard.app.domain.DeviceEntity;
+import com.switchboard.app.exceptions.DeviceNotFoundException;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.springframework.http.ResponseEntity;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.when;
+
+class DecoderControllerTest {
+
+    @InjectMocks
+    DecoderController decoderController;
+
+    @Mock
+    DecoderDaoImpl decoderService;
+
+    @Mock
+    DeviceDaoImpl deviceService;
+
+    //stubbed Objects
+    private DeviceEntity device1;
+    private DecoderEntity decoder1;
+
+    @BeforeEach
+    void setUp() throws Exception{
+        MockitoAnnotations.initMocks(this); //to be able to initiate decoderController object
+
+        //stubbing device and decoder objects
+        device1 = new DeviceEntity("1","Decoder #1","Running",null,null);
+        decoder1 = new DecoderEntity("1", device1);
+    }
+
+    @Test
+    final void testRetrieveAllDecoders(){
+        //stubbing DeviceEntity and DecoderEntity objects
+        DeviceEntity device2 = new DeviceEntity("2","Decoder #2","Failing",null,null);
+        DecoderEntity decoder2 = new DecoderEntity("2", device2);
+
+        //Adding stubbed objects to the list that should be returned when getDecoders is called
+        List<DecoderEntity> listOfDecoders= new ArrayList<DecoderEntity>();
+        listOfDecoders.add(decoder1);
+        listOfDecoders.add(decoder2);
+
+        when(decoderService.getDecoders()).thenReturn(listOfDecoders);
+
+        List allDecoders = decoderController.retrieveAllDecoders();
+
+        assertFalse(allDecoders.isEmpty(),"allDecoders list is empty."); //check if an empty list was returned
+        assertIterableEquals(listOfDecoders, allDecoders,"listOfDecoders and allDecoders lists are not equal."); //check both lists contents
+    }
+
+    //When a decoder is available in the DB
+    @Test
+    final void testRetrieveDecoder(){
+        when(decoderService.findDecoder("1")).thenReturn(java.util.Optional.of(decoder1));
+
+        DecoderEntity actualDecoder = decoderController.retrieveDecoder("1").getContent();
+
+        assertNotNull(actualDecoder, "actualDecoder object is null.");
+        assertEquals(decoder1, actualDecoder, "expectedDecoder and actualDecoder objects are not equal.");
```

Comment:
Could we make the asserts on the specific properties of the objects? I think this assert tests the address of the two objects.

Reply:
Here's the assertEquals implementation:
![assertEqual Implementation](https://user-images.githubusercontent.com/39838955/96353242-3bd7bd80-1098-11eb-8b5f-997966bd1c9d.png)

It uses .equals method to compare two objects which is implemented for all entity classes using the lombok plugin.